### ISSUE TITLE
Feature request: Make direnv ask for verification of changes to watched files

### ISSUE THREAD

**Issue by daniels** [direnv/direnv#448]
## Issue description

When sourcing or running external files from .envrc, direnv will execute changed code in those files without verification. This is clearly documented, but I still think it would still be great if the `direnv allow` process could be applied to such files as well.

Since direnv already knows how to watch other files and reload when they change, the same list of files would be a great candidate for files to also protect from unverified changes.

*Note: This is a very humble suggestion without any insight into how complex the implentation would be. Since the documentation mentions the lack of this feature, I guess it has been considered but deemed non-trivial.*

### Steps to reproduce

1. Source and watch an external file from .envrc
2. Call `direnv allow`
3. Change the external file

Direnv will reload and source the changed external file without verification.

**Comment by zimbatm** [comment]
Thinking about this a bit more: because direnv is not declarative, it wouldn't be possible for direnv to collect the list up-front. It means that in terms of user-experience, the only thing that it could do is interrupt the evaluation (for example `source_env` might be augmented to do that) to ask for more permission. In terms of UX, it wouldn't be the best experience I think.

**Comment by zimbatm** [comment]
Maybe the best solution would be to go the other way: a user would be asked to confirm for a certain directory to load the `.envrc` the first time direnv encounters that directory. And make it extra clear but dumping a big warning message that the whole directory is not permitted to load arbitrary code.

**Comment by daniels** [comment]
> Thinking about this a bit more: because direnv is not declarative, it wouldn't be possible for direnv to collect the list up-front. It means that in terms of user-experience, the only thing that it could do is interrupt the evaluation (for example `source_env` might be augmented to do that) to ask for more permission. In terms of UX, it wouldn't be the best experience I think.

Hm, could it use the watched files from the *previous* run as the list to check?

If .envrc is changed to source and watch a new file, that file will be unknown and unverified on first run, but the change to .envrc will be detected and the user could be expected to review any new includes as well. But if a watched file is removed from the .envrc *and* changed (or deleted) it would of course interrupt the user with a false alarm. But then again .envrc would also be changed so you would be interrupted anyway. If there was a command to allow all the changed files at once, I don't think it would not be very concerning.

But maybe I'm missing some more complex use case. For me the external file is most often a single `shell.nix` file that is sourced via "use nix", but with nested includes or inherited .envrc structures maybe the above would fail?

**Comment by timokau** [comment]
I agree that his would be a good idea. The way it is now `direnv allow` seems very inconsistent: every time I modify the envrc I need to approve it, but every time I modify the equally "risky" shell.nix, I don't.

@daniels solution sounds good, although there could be edge-cases where a file is imported conditionally. Another option would be to allow users to declaratively provide a list of files whose changes need to be approved.
